src/charm.py

```diff
@@ -18,98 +18,63 @@
-            "Changing permissions on existing content: {} -> {}".format(
-                existing_perms, perms
-            )
-        )
-        os.chmod(path, perms)
-
+def _write_file(path: str, content: bytes, perms: int) -> None:
+    """
+    Create or overwrite a file with the contents of a byte string.
 
-def install_ca_cert(ca_cert, name=None):
+    :param str path: The file path where the content should be written.
+    :param bytes content: The content to write to the file.
+    :param int perms: The file permissions to set on the created file.
     """
-    Install the given cert as a trusted CA.
+    uid = pwd.getpwnam("root").pw_uid
+    gid = grp.getgrnam("root").gr_gid
 
-    The ``name`` is the stem of the filename where the cert is written, and if
-    not provided, it will default to ``juju-{charm_name}``.
+    logger.debug(
+        "Writing file {} with root ownership and permissions {:o}".format(path, perms)
+    )
 
-    If the cert is empty or None, or is unchanged, nothing is done.
+    # Create a temporary file and replace the target file
+    # This ensures atomic file writing for safety
+    dir_name = os.path.dirname(path)
+    with tempfile.NamedTemporaryFile(dir=dir_name, delete=False) as tmp_file:
+        tmp_file.write(content)
+        tmp_file.flush()
+        os.fchown(tmp_file.fileno(), uid, gid)
+        os.fchmod(tmp_file.fileno(), perms)
+        temp_path = tmp_file.name
+
+    os.rename(temp_path, path)
+
+
+def _install_ca_cert(ca_cert: bytes):
     """
-    if not ca_cert:
-        return
-    if not isinstance(ca_cert, bytes):
-        ca_cert = ca_cert.encode("utf8")
-    # if not name:
-    #     name = 'juju-{}'.format(charm_name())
-    cert_file = "/usr/local/share/ca-certificates/{}.crt".format(name)
-    new_hash = hashlib.md5(ca_cert).hexdigest()
-    if file_hash(cert_file) == new_hash:
-        return
-    logger.info("Installing new CA cert at: {}".format(cert_file))
-    write_file(cert_file, ca_cert)
-    subprocess.check_call(["update-ca-certificates", "--fresh"])
-
-
-def file_hash(path, hash_type="md5"):
-    """Generate a hash checksum of the contents of 'path' or None if not found.
-
-    :param str hash_type: Any hash alrgorithm supported by :mod:`hashlib`,
-                        such as md5, sha1, sha256, sha512, etc.
+    Install the given cert as a trusted CA.
+
+    :param bytes ca_cert: The base64 decoded CA certificate.
     """
-    if os.path.exists(path):
-        h = getattr(hashlib, hash_type)()
-        with open(path, "rb") as source:
-            h.update(source.read())
-        return h.hexdigest()
-    else:
-        return None
+    cert_file = CA_CERT_PATH
+    logger.info("Installing new CA cert at: %s", cert_file)
+    _write_file(cert_file, ca_cert, 0o444)
+
+    # Execute the command to update the CA certificates
+    try:
+        subprocess.check_call(["update-ca-certificates", "--fresh"])
+    except subprocess.CalledProcessError as error:
```

Comment:
**nitpick:** 

What is the expect behavior for charm here?
Get into error status so user can use `juju resolve` to re-try? Or any clue/helping-message for user to know what's happening?

It's a little bit danger for me here because it's failing but charm do not change it's status to warn the user.

Reply:
Would it be okay to just have a helping message here? Block the whole charm for ca-certificate sounds a bit excessive.